src/controller.js

```diff
@@ -0,0 +1,601 @@
+   * the passed in ones.  This is a sanity check.
+   */
+  __getResponseProperty(doc, container, property, bucket, key) {
+    if (doc.root().name() !== container) {
+      throw new Error('Document does not have ' + container);
+    }
+
+    let lookingfor, foundBucket, foundKey;
+    
+    for (let child of doc.root().childNodes()) {
+      switch (child.name()) {
+        case 'Bucket':
+          foundBucket = child.text();
+          break;
+        case 'Key':
+          foundKey = child.text();
+          break;
+        case property:
+          lookingfor = child.text();
+          break;
+      }
+    }
+
+    if (foundBucket !== bucket) {
+      throw new Error('Document contains incorrect Bucket');
+    }
+
+    if (foundKey !== key) {
+      throw new Error('Document contains incorrect Key');
+    }
+
+    if (typeof lookingfor === 'undefined') {
+      throw new Error('Document does not contain ' + property);
+    }
+
+    return lookingfor;
+  }
+  /**
+   * Obtain an UploadId from an Initiate Multipart Upload response body
+   */
+  __getUploadId(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'InitiateMultipartUploadResult', 'UploadId', bucket, key);
+  }
+
+  /**
+   * Obtain the ETag for the committed multipart upload
+   */
+  __getMultipartEtag(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'CompleteMultipartUploadResult', 'ETag', bucket, key);
+  }
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
+    permissions = runSchema(permissions, schemas.permissions.required());
+    let {acl, read, write, readAcp, writeAcp, fullControl} = permissions;
+    if (acl) {
+      return [['x-amz-acl', permissions.acl]];
+    }
+
+    let perms = [];
+
+    if (read) {
+      perms.push(['x-amz-grant-read', read]);
+    }
+
+    if (write) {
+      perms.push(['x-amz-grant-write', write]);
+    }
+
+    if (readAcp) {
+      perms.push(['x-amz-grant-read-acp', readAcp]);
+    }
+
+    if (writeAcp) {
+      perms.push(['x-amz-grant-write-acp', writeAcp]);
+    }
+
+    if (fullControl) {
+      perms.push(['x-amz-grant-full-control', fullControl]);
+    }
+
+    return perms;
+  }
+
+  __s3hostname(bucket) {
+    let hostname = `${bucket}.${this.s3host}`;
+    if (this.s3port) {
+      hostname += ':' + this.s3port;
+    }
+    return hostname;
+  }
+ 
+  /**
+   * Initiate a Multipart upload and return the UploadIp that
+   * Amazon has assigned for this multipart upload
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html
+   */
+  async initiateMultipartUpload(opts) {
+
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.mpSize.required(),
+      permissions: schemas.permissions,
```

Comment:
Ahh yes, I should add the ability to pass metadata into the methods.  Good catch.  I'll look into what's required to add storage class.